x/stake/keeper/distribute.go

```diff
@@ -136,10 +144,48 @@ func (k Keeper) DistributeInBreathBlock(ctx sdk.Context, sideChainId string) sdk
 	var toSaveValDistAddrs []types.StoredValDistAddr // mapping between validator and distribution address, to be saved
 
 	bondDenom := k.BondDenom(ctx)
+	feeFromBscToBcRatio := k.FeeFromBscToBcRatio(ctx)
+	avgFeeForBcVals := sdk.ZeroDec()
+	if sdk.IsUpgrade(sdk.BEPHHH) && sideChainId == types.MockSideChainIDForBeaconChain {
+		feeForAllBcValsCoins := k.BankKeeper.GetCoins(ctx, FeeForAllBcValsAccAddr)
+		feeForAllBcVals := feeForAllBcValsCoins.AmountOf(bondDenom)
+		avgFeeForBcVals = sdk.NewDec(feeForAllBcVals / int64(len(validators)))
+	}
+	ctx.Logger().Debug("DistributeInBreathBlock", "height", ctx.BlockHeight(), "sideChainId", sideChainId, "avgFeeForBcVals", avgFeeForBcVals)
+
 	for _, validator := range validators {
 		distAccCoins := k.BankKeeper.GetCoins(ctx, validator.DistributionAddr)
+		//ctx.Logger().Debug("DistributeInBreathBlock", "distAccCoins", distAccCoins)
 		totalReward := distAccCoins.AmountOf(bondDenom)
-		totalRewardDec := sdk.ZeroDec()
+		totalRewardDec := sdk.NewDec(totalReward)
+		if sdk.IsUpgrade(sdk.BEPHHH) {
+			if sideChainId != types.MockSideChainIDForBeaconChain {
+				// split a portion of fees to BC validators
+				feeToBC := totalRewardDec.Mul(feeFromBscToBcRatio)
+				if feeToBC.RawInt() > 0 {
+					if _, _, err := k.BankKeeper.SubtractCoins(ctx, validator.DistributionAddr, sdk.Coins{sdk.NewCoin(bondDenom, feeToBC.RawInt())}); err != nil {
+						panic(err)
```

Comment:
We can do `BankKeeper.SetCoins` here than `sub` and `add`

Reply:
`BankKeeper.SetCoins` will overwrite the coins of an account directly, so we have to get the original coins, do the calculation, and then set it. It's exactly what `SubtractCoins` and `AddCoins` functions do. I don't get why it's better.

```
func setCoins(ctx sdk.Context, am auth.AccountKeeper, addr sdk.AccAddress, amt sdk.Coins) sdk.Error {
	acc := am.GetAccount(ctx, addr)
	if acc == nil {
		acc = am.NewAccountWithAddress(ctx, addr)
	}
	err := acc.SetCoins(amt)
	if err != nil {
		// Handle w/ #870
		panic(err)
	}
	am.SetAccount(ctx, acc)
	return nil
}

func subtractCoins(ctx sdk.Context, am auth.AccountKeeper, addr sdk.AccAddress, amt sdk.Coins) (sdk.Coins, sdk.Tags, sdk.Error) {
	oldCoins := getCoins(ctx, am, addr)
	newCoins := oldCoins.Minus(amt)
	if !newCoins.IsNotNegative() {
		return amt, nil, sdk.ErrInsufficientCoins(fmt.Sprintf("%s < %s", oldCoins, amt))
	}
	err := setCoins(ctx, am, addr, newCoins)
	tags := sdk.NewTags("sender", []byte(addr.String()))
	return newCoins, tags, err

```

Reply:
sorry, typo, I mean `SendCoins`

Reply:
fixed